### PR TITLE
feat(sdk): alias v2 imports and provide DeprecationWarning

### PR DESCRIPTION
**Pull request by connor-mccarthy**
**Description of your changes:**
These changes modifies the `kfp.v2` namespace such it imports from the main `kfp` namespace. Put simply: it makes old `v2` imports backward compatible.

Solution:
To do this in a _fully_-backward compatible way, we needed to replicate the module structure in the `v2/` directory.

Alternatives:
1. Putting `from kfp import *` in `sdk/v2/__init__.py`. This does not provide the expected behavior. This only imports objects in `kfp/__init__.py`, which include only `TYPE_CHECK` and `COMPILING_FOR_V2`. Consequently, `from kfp.v2 import dsl` does not work, since `dsl` is never actually imported to `kfp.v2` from the `kfp`.
2. Putting `from kfp import dsl, compiler, components`  in `sdk/v2/__init__.py`. This gets us closer, because it enables `from kfp.v2 import dsl`, but does not enable `kfp.v2.dsl import pipeline`, because the `.dsl` module does not exist under `kfp/v2/`.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/v2/v2_namespace_alias_test.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import unittest
+
+
+class V2NamespaceAliasTest(unittest.TestCase):
+    """Test that imports of both modules and objects are aliased
+    (e.g. all import path variants work).
+    """
+
+    def test_1_import_module(self):  # pylint: disable=no-self-use
+        from kfp.v2 import dsl as dsl2
```
Comment (ji-yaqi) [inline sdk/python/kfp/v2/v2_namespace_alias_test.py:25]: Maybe you can add the check warnings here, check whether warnings are right. 

And for usage, we normally use:
```
from kfp.v2 import dsl

@dsl.pipeline(name="xxx")
def pipeline:
etc.
``` 
Try to put a real usage here, as we sometimes refer to these usages from tests.

File: sdk/python/kfp/v2/v2_namespace_alias_test.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    def test_2_import_object(self):  # pylint: disable=no-self-use
+        from kfp.v2.dsl import pipeline as pipeline2
```
Comment (ji-yaqi) [inline sdk/python/kfp/v2/v2_namespace_alias_test.py:28]: Similar here, example would be 
```
from kfp.v2.dsl import Dataset

@dsl.pipeline(name="xxx")
def pipeline:
etc.
```

File: sdk/python/kfp/v2/components.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# Copyright 2020 The Kubeflow Authors
```
Comment (ji-yaqi) [inline sdk/python/kfp/v2/components.py:1]: nit: change these to 2022 :)

Comment (connor-mccarthy) [reply sdk/python/kfp/v2/components.py:1]: Just did this in a separate PR: https://github.com/kubeflow/pipelines/pull/7401

File: sdk/python/kfp/v2/v2_namespace_alias_test.py (hunk 1 above)
Comment (connor-mccarthy) [reply sdk/python/kfp/v2/v2_namespace_alias_test.py]: Good idea on adding usage. Done.

I worked on checking that the warning is raised using `self.assertWarns` but it turned out to be fairly challenging to implement reliably and probably not worth the effort, I think. The issue is that `unittest` appears to use the same execution session for all tests, so it's not certain that the `kfp.v2` path will be hit for the first time in the  `sdk/python/kfp/v2/v2_namespace_alias_test.py` file (in fact, this has never happened for me in practice). Once the `kfp.v2` module is loaded (by another unrelated test, in this case), the `DeprecationWarning` is not raised. I tried a few workarounds using `sys.modules`, but this was unsuccessful.

File: sdk/python/kfp/v2/v2_namespace_alias_test.py (hunk 2 above)
Comment (connor-mccarthy) [reply sdk/python/kfp/v2/v2_namespace_alias_test.py]: Added usage.

File: sdk/python/kfp/v2/v2_namespace_alias_test.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
+# Copyright 2022 The Kubeflow Authors
```
Comment (chensun) [inline sdk/python/kfp/v2/v2_namespace_alias_test.py:1]: Feels a bit odds to have a test file hanging out in the v2 folder. How about merging it into https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/compiler/compiler_test.py
WDYT?

File: sdk/python/kfp/v2/v2_namespace_alias_test.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        with tempfile.TemporaryDirectory() as tempdir:
+            # you can e.g. create a file here:
+            temp_filepath = os.path.join(tempdir, 'hello_world_pipeline.json')
+            compiler.Compiler().compile(
+                pipeline_func=pipeline_hello_world, package_path=temp_filepath)
+
+            with open(temp_filepath, "r") as f:
+                ir_template = json.load(f)
+
+        self.assertIsInstance(ir_template, dict)
```
Comment (chensun) [inline sdk/python/kfp/v2/v2_namespace_alias_test.py:52]: I guess if `json.load(f)` didn't raise any error, this must be true? So seems unnecessary check?
On the other hand, can we assert the v2 aliasing warning was throw?

Comment (chensun) [inline sdk/python/kfp/v2/v2_namespace_alias_test.py:52]: I guess if `json.load(f)` didn't raise any error, this must be true? So seems unnecessary check?
On the other hand, can we assert the v2 aliasing warning was throw?

Comment (connor-mccarthy) [reply sdk/python/kfp/v2/v2_namespace_alias_test.py:52]: Agreed -- I'll remove the test.

I worked on asserting that the warning was thrown for some time, but found it to be very challenging and perhaps not worth the effort. My explanation is here: https://github.com/kubeflow/pipelines/pull/7398#discussion_r824233150

Comment (connor-mccarthy) [reply sdk/python/kfp/v2/v2_namespace_alias_test.py]: FYI, I'm trying one new approach for remedying the `assertWarns` issue mentioned above. Will follow up.

File: sdk/python/kfp/v2/__init__.py [hunk 6]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2020 The Kubeflow Authors
+# Copyright 2022 The Kubeflow Authors
```
Comment (chensun) [inline sdk/python/kfp/v2/__init__.py:1]: nit: 2020-2022

File: sdk/python/kfp/v2/v2_namespace_alias_test.py (hunk 4 above)
Comment (chensun) [reply sdk/python/kfp/v2/v2_namespace_alias_test.py:1]: You're right that neither follows the convention. Putting this in compiler_test.py might be cleaner, but I don't have a strong opinion on this. So up to you, I'm fine keeping it here :)

File: sdk/python/kfp/v2/v2_namespace_alias_test.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+    @classmethod
+    def setUp(cls):
+        # need to unload the kfp.v2 module before each test to ensure that the Deprecation warning is raised each time
+        keys = {k for k, _ in sys.modules.items() if k.startswith("kfp")}
+        for key in keys:
+            del sys.modules[key]
+
+    def test_import_namespace(self):  # pylint: disable=no-self-use
+        with self.assertWarns(DeprecationWarning):
```
Comment (chensun) [inline sdk/python/kfp/v2/v2_namespace_alias_test.py:36]: There might be other deprecation warnings. Consider using `assertWarnsRegex`, which allows you to (partially) match the warning message. 

e.g.: https://github.com/kubeflow/pipelines/blob/b693909a0879d478c4cb8711932b06c956629ae0/sdk/python/kfp/deprecated/components_tests/test_components.py#L1106-L1110

**Comment by connor-mccarthy** [comment]
@chensun, please see my [code comment](https://github.com/kubeflow/pipelines/blob/c28d5a884e3aea674d1067dccd48e7314382751e/sdk/python/kfp/compiler/compiler_test.py#L579-L581) for my conclusion regarding use of `self.assertWarnsRegex`.

**Comment by chensun** [comment]
/lgtm
/approve

Thanks!

**Pull request merged**